From 7c66bf98f6e6cd88fd5c9f4d027f19e531f1643f Mon Sep 17 00:00:00 2001
From: Manikanta Mylavarapu <mmanikan@qti.qualcomm.com>
Date: Thu, 29 Jan 2026 14:54:39 +0530
Subject: [PATCH] pcie : qcom: enable Q6 RDDM transistion for link recovery

Add support in the Qualcomm DWC PCIe host driver to move the
Q6 subsystem to RDDM when a PCIe link-down event occurs.
Entering RDDM allows Q6 to collect crash information and helps
identify the cause of the link failure.

During link recovery, the host asserts PERST# to trigger the Q6
transition to RDDM. The driver then waits for either a wake
interrupt or a timeout on platforms without wake IRQ support.
Client drivers will perform endpoint recovery once Q6 enters RDDM.

This link-recovery flow is enabled only when AER support is present,
since AER provides the necessary error-handling infrastructure.

Once the upstream patch is merged, the AER-specific conditional
checks will be removed, and the relevant functionality will be
validated with both AER enabled and disabled.

Change-Id: I653ebdfa0a38fb0e59062dcc413ae55bceafe68c
Signed-off-by: Manikanta Mylavarapu <mmanikan@qti.qualcomm.com>
---
 drivers/pci/controller/dwc/pcie-qcom.c | 68 +++++++++++++++++++++++++-
 1 file changed, 67 insertions(+), 1 deletion(-)

diff --git a/drivers/pci/controller/dwc/pcie-qcom.c b/drivers/pci/controller/dwc/pcie-qcom.c
index ebda1ecb97ba..ebbefa9e0f07 100644
--- a/drivers/pci/controller/dwc/pcie-qcom.c
+++ b/drivers/pci/controller/dwc/pcie-qcom.c
@@ -350,6 +350,10 @@ struct qcom_pcie {
 	uint32_t domain;
 	uint32_t num_lanes;
 	int global_irq;
+#if IS_ENABLED(CONFIG_PCIEAER)
+	int wake_irq;
+	struct completion wake_event;
+#endif
 	bool enable_vc;
 	bool enable_iatu;
 	u32 iatu_ib0_base_addr[3];	/* Low , High, Limit */
@@ -367,8 +371,12 @@ struct qcom_pcie_info{
 LIST_HEAD(qcom_pcie_list);
 
 #define to_qcom_pcie(x)		dev_get_drvdata((x)->dev)
+
+#if IS_ENABLED(CONFIG_PCIEAER)
+#define WAKE_EVENT_MSEC_DELAY 5000
 static int qcom_pcie_reset_slot(struct pci_host_bridge *bridge,
 				struct pci_dev *pdev);
+#endif
 
 static void qcom_ep_reset_assert(struct qcom_pcie *pcie)
 {
@@ -395,10 +403,23 @@ static int qcom_pcie_start_link(struct dw_pcie *pci)
 	return 0;
 }
 
+#if IS_ENABLED(CONFIG_PCIEAER)
+static irqreturn_t qcom_pcie_wake_irq_handler(int irq, void *data)
+{
+	struct qcom_pcie *pcie = data;
+
+	complete(&pcie->wake_event);
+
+	return IRQ_HANDLED;
+}
+#endif
+
 static irqreturn_t qcom_pcie_global_irq_thread_fn(int irq, void *data)
 {
 	struct qcom_pcie *pcie = data;
+#if IS_ENABLED(CONFIG_PCIEAER)
 	struct dw_pcie_rp *pp = &pcie->pci->pp;
+#endif
 	u32 status, mask;
 
 	status = readl_relaxed(pcie->parf + PARF_INT_ALL_STATUS);
@@ -410,7 +431,17 @@ static irqreturn_t qcom_pcie_global_irq_thread_fn(int irq, void *data)
 
 	if (status & PARF_INT_ALL_LINK_DOWN) {
 		dev_info(pcie->pci->dev, "Received Link down event\n");
+		/* We should skip pci_host_handle_link_down() for legacy devices
+		 * where AER is not enabled. Else it reset's the secondary or
+		 * sub-ordinate bus, subsequently EP get reset.
+		 * Hence It should be avoided.
+		 *
+		 * To Do: After the upstream patch is merged, remove the AER check
+		 * and verify the behavior with both AER enabled and disabled.
+		 */
+#if IS_ENABLED(CONFIG_PCIEAER)
 		pci_host_handle_link_down(pp->bridge);
+#endif
 	} else if (status & PARF_INT_ALL_LINK_UP) {
 		dev_info(pcie->pci->dev, "Received Link up event\n");
 	}
@@ -1458,8 +1489,16 @@ static int qcom_pcie_host_init(struct dw_pcie_rp *pp)
 			goto err_assert_reset;
 	}
 
+	/* reset_slot() callback is required only if AER enabled.
+	 * For non AER case, reset_slot() may reset the endpoint
+	 * if user executes "echo 1 > /sys/bus/pci/devices/<device>/reset"
+	 *
+	 * To Do: After the upstream patch is merged, remove the AER check
+	 * and verify the behavior with both AER enabled and disabled.
+	 */
+#if IS_ENABLED(CONFIG_PCIEAER)
 	pp->bridge->reset_slot = qcom_pcie_reset_slot;
-
+#endif
 	return 0;
 
 err_assert_reset:
@@ -1840,6 +1879,7 @@ static void qcom_pcie_init_debugfs(struct qcom_pcie *pcie)
 				    qcom_pcie_link_transition_count);
 }
 
+#if IS_ENABLED(CONFIG_PCIEAER)
 static int qcom_pcie_reset_slot(struct pci_host_bridge *bridge,
 				struct pci_dev *pdev)
 {
@@ -1851,6 +1891,15 @@ static int qcom_pcie_reset_slot(struct pci_host_bridge *bridge,
 	u32 val;
 	int ret;
 
+	/* assert perst to move to RDDM state */
+	qcom_ep_reset_assert(pcie);
+
+	/* Wait for wake IRQ event or WAKE_EVENT_MSEC_DELAY timeout */
+	wait_for_completion_timeout(&pcie->wake_event, msecs_to_jiffies(WAKE_EVENT_MSEC_DELAY));
+
+	/* Reset completion for next use */
+	reinit_completion(&pcie->wake_event);
+
 	/* Wait for the pending transactions to be completed */
 	ret = readl_relaxed_poll_timeout(pcie->parf + PARF_STATUS, val,
 					 val & FLUSH_COMPLETED, 10,
@@ -1906,6 +1955,7 @@ static int qcom_pcie_reset_slot(struct pci_host_bridge *bridge,
 
 	return ret;
 }
+#endif
 
 int pcie_rescan(int val)
 {
@@ -2211,6 +2261,22 @@ static int __qcom_pcie_probe(struct platform_device *pdev,
 		}
 	}
 
+#if IS_ENABLED(CONFIG_PCIEAER)
+	init_completion(&pcie->wake_event);
+	pcie->wake_irq = platform_get_irq_byname_optional(pdev, "wake_irq");
+	if (pcie->wake_irq >= 0) {
+		ret = devm_request_irq(&pdev->dev, pcie->wake_irq,
+				       qcom_pcie_wake_irq_handler,
+				       IRQF_TRIGGER_FALLING,
+				       "pcie-wake", pcie);
+		if (ret) {
+			dev_err(&pdev->dev, "Unable to request wake irq\n");
+			pm_runtime_disable(&pdev->dev);
+			goto err_phy_exit;
+		}
+	}
+#endif
+
 	if (!rc_idx) {
 		ret = bus_create_file(&pci_bus_type, &bus_attr_slot_rescan);
 		if (ret != 0) {
-- 
2.34.1

